File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
import sys

input = sys.stdin.readline
N, M = map(int, input().split())
trees = list(map(int, input().split()))

left, right = 0, max(trees)
while left <= right:
    mid = (left + right) // 2
    current_length = 0
    for tree in trees:
        if tree > mid:
            current_length += tree - mid

    if current_length >= M:
        left = mid + 1
    else:
        right = mid - 1

print(right)

File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

class Main {
    static int N, M;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        int[] tree = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        Arrays.sort(tree);

        int left = 0, right = tree[N - 1], answer = 0;
        while (left < right) {
            int mid = (left + right) / 2;
            long cnt = 0;
            for (int i : tree) {
                if (i > mid) {
                    cnt += i - mid;
                }
            }

            if (cnt > M) {
                answer = mid;
                left = mid + 1;
            } else if (cnt < M) {
                right = mid;
            } else {
                answer = mid;
                break;
            }
        }
        System.out.println(answer);
    }
}